File: src/background/heartbeat.ts
import { existsSync, writeFileSync } from "node:fs";
import { resolve } from "node:path";
import { createLogger } from "../logger";
import { runAgent } from "../agent";
import { getOutbox } from "./outbox";

const log = createLogger("background:heartbeat");

const HEARTBEAT_PATH = resolve(process.cwd(), "user-space", "memory", "HEARTBEAT.md");
const DEFAULT_INTERVAL_MS = 30 * 60 * 1000; // 30 minutes
const SESSION_PEER_ID = "default";

/**
 * HeartbeatService — periodically reads HEARTBEAT.md and dispatches
 * each unchecked "- [ ] task" line as an agent instruction.
 *
 * The agent itself can modify HEARTBEAT.md (via write_file) to schedule
 * future self-directed thinking.
 */
export class HeartbeatService {
  private timer: ReturnType<typeof setInterval> | null = null;
  private isTicking = false;
  private target: { channel: "webchat" | "feishu"; peerId: string };

  constructor(target: { channel: "webchat" | "feishu"; peerId: string } = {
    channel: "webchat",
    peerId: "broadcast",
  }) {
    this.target = target;
  }

  start(intervalMs = DEFAULT_INTERVAL_MS): void {
    this.ensureHeartbeatFile();
    this.timer = setInterval(() => this.tick(), intervalMs);
    log.info("HeartbeatService started", { intervalMs });
  }

  stop(): void {
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
    log.info("HeartbeatService stopped");
  }

  /** Exposed for manual triggers and testing. */
  async tick(): Promise<void> {
    if (this.isTicking) {
      log.debug("Heartbeat tick skipped — previous tick still running");
      return;
    }
    this.isTicking = true;

    try {
      const tasks = await this.readTasks();
      if (tasks.length === 0) {
        log.debug("Heartbeat tick: no tasks");
        return;
      }

      log.info("Heartbeat tick", { tasks: tasks.length });

      for (const task of tasks) {
        try {
          const chunks: string[] = [];
          for await (const event of runAgent("heartbeat", SESSION_PEER_ID, task)) {
            if (event.type === "text") chunks.push(event.content ?? "");
          }
          const output = chunks.join("").trim();
          if (output) {
            await getOutbox().push("heartbeat", this.target, output);
          }
        } catch (err) {
          log.error("Heartbeat task failed", {
            task: task.slice(0, 80),
            error: err instanceof Error ? err.message : String(err),
          });
        }
      }
    } finally {
      this.isTicking = false;
    }
  }

  private async readTasks(): Promise<string[]> {
    if (!existsSync(HEARTBEAT_PATH)) return [];
    const text = await Bun.file(HEARTBEAT_PATH).text();
    const tasks: string[] = [];
    for (const line of text.split("\n")) {
      const m = line.match(/^-\s+\[\s+\]\s+(.+)/);
      if (m) tasks.push(m[1].trim());
    }
    return tasks;
  }

  private ensureHeartbeatFile(): void {
    if (!existsSync(HEARTBEAT_PATH)) {
      const template = [
        "# Heartbeat Tasks",
        "",
        "Add tasks below using `- [ ] description`.",
        "The agent checks this file every 30 minutes and executes each unchecked item.",
        "The agent can also add tasks here to self-schedule future thinking.",
        "",
        "## Examples (remove or edit these)",
        "",
        "- [ ] Review any reminders or pending items",
      ].join("\n") + "\n";
      writeFileSync(HEARTBEAT_PATH, template);
      log.info("HEARTBEAT.md created at", { path: HEARTBEAT_PATH });
    }
  }
}

let _heartbeat: HeartbeatService | null = null;

export function getHeartbeatService(): HeartbeatService {
  if (!_heartbeat) _heartbeat = new HeartbeatService();
  return _heartbeat;
}

File: src/background/heartbeat.py
"""Heartbeat service — periodic task execution from HEARTBEAT.md, matching TS background/heartbeat.ts."""

from __future__ import annotations

import asyncio
import re
from pathlib import Path

from src.utils.logger import create_logger

log = create_logger("background:heartbeat")

HEARTBEAT_PATH = Path("user-space/memory/HEARTBEAT.md")
DEFAULT_INTERVAL_S = 30 * 60  # 30 minutes
UNCHECKED_RE = re.compile(r"^-\s+\[\s+\]\s*(.*)", re.MULTILINE)


class HeartbeatService:
    def __init__(self, interval_s: float = DEFAULT_INTERVAL_S) -> None:
        self._interval = interval_s
        self._task: asyncio.Task | None = None
        self._ticking = False

    def start(self) -> None:
        if self._task:
            return
        self._task = asyncio.get_event_loop().create_task(self._loop())
        log.info("Heartbeat service started", {"intervalMin": self._interval / 60})

    def stop(self) -> None:
        if self._task:
            self._task.cancel()
            self._task = None

    async def tick(self) -> None:
        if self._ticking:
            log.debug("Heartbeat tick already in progress, skipping")
            return
        self._ticking = True
        try:
            await self._execute_tasks()
        finally:
            self._ticking = False

    async def _loop(self) -> None:
        while True:
            await asyncio.sleep(self._interval)
            try:
                await self.tick()
            except Exception as e:
                log.warn("Heartbeat tick error", {"error": str(e)})

    async def _execute_tasks(self) -> None:
        if not HEARTBEAT_PATH.exists():
            return

        content = HEARTBEAT_PATH.read_text(encoding="utf-8")
        tasks = UNCHECKED_RE.findall(content)

        if not tasks:
            return

        log.info("Heartbeat executing tasks", {"count": len(tasks)})

        from src.agents.init import run_agent
        from src.background.outbox import get_outbox

        for task_text in tasks:
            try:
                text_parts: list[str] = []
                async for event in run_agent("heartbeat", "default", task_text):
                    if event.type == "text" and event.content:
                        text_parts.append(event.content)

                result = "".join(text_parts)
                if result:
                    await get_outbox().push(
                        "heartbeat",
                        {"channel": "webchat", "peerId": "broadcast"},
                        result,
                    )
            except Exception as e:
                log.warn("Heartbeat task failed", {"task": task_text[:50], "error": str(e)})


_service: HeartbeatService | None = None


def get_heartbeat_service() -> HeartbeatService:
    global _service
    if _service is None:
        _service = HeartbeatService()
    return _service
